channelmonitor/channelmonitor.go

```diff
@@ -321,52 +328,94 @@ func (mc *monitoredChannel) isRestarting() bool {
 			mc.restartedAt = time.Now()
-			mc.consecutiveRestarts++
-			restartCount = mc.consecutiveRestarts
+		} else {
+			// There is already a restart in progress, so queue up a restart
+			// for after the current one has completed
+			mc.restartQueued = true
 		}
 	}
 	mc.restartLk.Unlock()
 
 	// Check if channel is already being restarted
 	if !restartedAt.IsZero() {
-		log.Debugf("%s: restart called but already restarting channel (for %s so far; restart backoff is %s)",
+		log.Infof("%s: restart called but already restarting channel, "+
+			"waiting to restart again (since %s; restart backoff is %s)",
 			mc.chid, time.Since(restartedAt), mc.cfg.RestartBackoff)
 		return
 	}
 
+	for {
+		// Send the restart message
+		err := mc.doRestartChannel()
+		if err != nil {
+			// If there was an error restarting, close the channel and shutdown
+			// the monitor
+			mc.closeChannelAndShutdown(err)
+			return
+		}
+
+		// Restart has completed, check if there's another restart queued up
+		restartAgain := false
+		mc.restartLk.Lock()
+		{
+			if mc.restartQueued {
+				// There's another restart queued up, so reset the restart time
+				// to now
+				mc.restartedAt = time.Now()
+				restartAgain = true
+				mc.restartQueued = false
+			} else {
+				// No other restarts queued up, so clear the restart time
+				mc.restartedAt = time.Time{}
+			}
+		}
+		mc.restartLk.Unlock()
+
+		if !restartAgain {
+			// No restart queued, we're done
+			return
+		}
+
+		// There was a restart queued, restart again
+		log.Infof("%s: restart was queued - restarting again", mc.chid)
+	}
+}
+
+func (mc *monitoredChannel) doRestartChannel() error {
+	// Keep track of the number of consecutive restarts with no data
+	// transferred
+	mc.restartLk.Lock()
+	mc.consecutiveRestarts++
+	restartCount := mc.consecutiveRestarts
+	mc.restartLk.Unlock()
+
 	if uint32(restartCount) > mc.cfg.MaxConsecutiveRestarts {
-		// If no data has been transferred since the last transfer, and we've
-		// reached the consecutive restart limit, close the channel and
-		// shutdown the monitor
-		err := xerrors.Errorf("%s: after %d consecutive restarts failed to transfer any data", mc.chid, restartCount)
-		mc.closeChannelAndShutdown(err)
-		return
+		// If no data has been transferred since the last restart, and we've
+		// reached the consecutive restart limit, return an error
+		return xerrors.Errorf("%s: after %d consecutive restarts failed to transfer any data", mc.chid, restartCount)
 	}
 
 	// Send the restart message
+	log.Infof("%s: restarting (%d consecutive restarts)", mc.chid, restartCount)
 	err := mc.sendRestartMessage(restartCount)
 	if err != nil {
-		// If the restart message could not be sent, close the channel and
-		// shutdown the monitor
-		mc.closeChannelAndShutdown(err)
-		return
+		log.Warnf("%s: restart failed, trying again: %s", mc.chid, err)
+		// If the restart message could not be sent, or there was a timeout
+		// waiting for the restart to be acknowledged, try again
```

Comment:
Shouldn't we have some kind of a sleep here? Seems like we can end up firing a lot of consecutive messages.

Reply:
`sendRestartMessage` tries to connect to the peer, and then waits for a restart ack.
When connecting to the peer, it has retries with backoff. If it returns an error it will be after having tried several times.
If it returns an error waiting for a restart ack it will be after a timeout.

So in both of these cases there is an implied delay before it returns an error.